core/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java

```diff
@@ -0,0 +1,259 @@
+                                        logger.trace("[{}] failed to create rollover index", t, updateRequest.index());
+                                    } else {
+                                        logger.debug("[{}] failed to create rollover index", t, updateRequest.index());
+                                    }
+                                    listener.onFailure(t);
+                                }
+                            });
+                        } else {
+                            indexAliasesService.indicesAliases(
+                                prepareIndicesAliasesRequest(sourceIndexName, rolloverIndexName, rolloverRequest),
+                                new IndicesAliasesListener(rolloverResponse, listener));
+                        }
+                    } else {
+                        // conditions not met
+                        listener.onResponse(
+                            new RolloverResponse(sourceIndexName, sourceIndexName, conditionStatus, false, false,
+                                false)
+                        );
+                    }
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    listener.onFailure(e);
+                }
+            }
+        );
+
+    }
+
+    private static final class IndicesAliasesListener implements ActionListener<ClusterStateUpdateResponse> {
+
+        private final ActionListener<RolloverResponse> listener;
+        private final RolloverResponse response;
+
+        public IndicesAliasesListener(RolloverResponse response, ActionListener<RolloverResponse> listener) {
+            this.response = response;
+            this.listener = listener;
+        }
+
+        @Override
+        public void onResponse(ClusterStateUpdateResponse clusterStateUpdateResponse) {
+            listener.onResponse(response);
+        }
+
+        @Override
+        public void onFailure(Throwable e) {
+            listener.onFailure(e);
+        }
+    }
+
+    static IndicesAliasesClusterStateUpdateRequest prepareIndicesAliasesRequest(String concreteSourceIndex, String targetIndex,
+                                                                                RolloverRequest request) {
+        final IndicesAliasesClusterStateUpdateRequest updateRequest = new IndicesAliasesClusterStateUpdateRequest()
+            .ackTimeout(request.ackTimeout())
+            .masterNodeTimeout(request.masterNodeTimeout());
+        AliasAction[] actions = new AliasAction[2];
+        actions[0] = new AliasAction(AliasAction.Type.ADD, targetIndex, request.getSourceAlias());
+        actions[1] = new AliasAction(AliasAction.Type.REMOVE, concreteSourceIndex, request.getSourceAlias());
+        updateRequest.actions(actions);
+        return updateRequest;
+    }
+
+    static String generateRolloverIndexName(String sourceIndexName) {
```

Comment:
I am not sure if I like this. I think we should be strict here and ensure that the index name matches with `.*(_\d)*^` and strip it off only at the end. If we hit `NFE` we should rethrow!